### PR TITLE
feat: Update wholesale event and calculate QuantityQuality

### PR DESCRIPTION
**Pull request by MWO1024**
## Description
Update the old CalculationResultCompleted event to the new EnergyResultProducedV2 event. The update includes a new mapping from a set of qualities to a single quality value.

### CODE REVIEW

File: source/BuildingBlocks.Domain/Models/QV2/EdiQuality.cs [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+namespace Energinet.DataHub.EDI.BuildingBlocks.Domain.Models.QV2;
+
+public enum EdiQuality
```
Comment (lasrinnil) [inline source/BuildingBlocks.Domain/Models/QV2/EdiQuality.cs:17]: please provide a summary if what this describes

Comment (lasrinnil) [inline source/BuildingBlocks.Domain/Models/QV2/EdiQuality.cs:17]:
```suggestion
public enum EdiQuantityQuality
```

File: source/BuildingBlocks.Domain/Models/EdiQuantityQuality.cs [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+namespace Energinet.DataHub.EDI.BuildingBlocks.Domain.Models;
+
+/// <summary>
+///     Represents the quantity quality of a data point in a time series.
+///     The enum is an internal representation of used solely internally in the the EDI system.
+///     This enum can and will be translated to an output format specific code and/or value as part of the generation of
+///     outgoing messages.
+/// </summary>
+public enum EdiQuantityQuality
```
Comment (lasrinnil) [inline source/BuildingBlocks.Domain/Models/EdiQuantityQuality.cs:23]: CalcualtedQuantiyQuality

File: source/BuildingBlocks.Domain/Models/EdiQuantityQuality.cs [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+namespace Energinet.DataHub.EDI.BuildingBlocks.Domain.Models;
+
+/// <summary>
+///     Represents the quantity quality of a data point in a time series.
+///     The enum is an internal representation of used solely internally in the the EDI system.
```
Comment (lasrinnil) [inline source/BuildingBlocks.Domain/Models/EdiQuantityQuality.cs:19]: make sure that we use the provided documentation for the enum weights - i.e: missing wins over the others 🗡️

File: source/IntegrationTests/Application/Transactions/Aggregations/WhenAnAggregationResultIsAvailableTests.cs [hunk 4]
```diff
@@ -78,14 +80,14 @@ await _gridAreaBuilder
             .StoreAsync(GetService<IMasterDataClient>());
 
         _eventBuilder
-            .WithProcessType(ProcessType.BalanceFixing)
+            .WithProcessType(CalculationType.BalanceFixing)
```
Comment (lasrinnil) [inline source/IntegrationTests/Application/Transactions/Aggregations/WhenAnAggregationResultIsAvailableTests.cs:83]:
```suggestion
            .WithCalculationType(CalculationType.BalanceFixing)
```

File: source/IntegrationTests/Infrastructure/InboxEvents/TestAggregatedTimeSeriesRequestAcceptedHandlerSpy.cs [hunk 5]
```diff
@@ -59,15 +59,16 @@ public Task Handle(AggregatedTimeSerieRequestWasAccepted notification, Cancellat
...
-            QuantityQuality.Measured => Quality.Measured.Name,
-            QuantityQuality.Missing => Quality.Missing.Name,
-            QuantityQuality.Estimated => Quality.Estimated.Name,
-            QuantityQuality.Unspecified => throw new InvalidOperationException("Quality is not specified"),
+            QuantityQuality.Incomplete => EdiQuantityQuality.NotAvailable,
+            QuantityQuality.Measured => EdiQuantityQuality.Measured,
+            QuantityQuality.Missing => EdiQuantityQuality.Missing,
+            QuantityQuality.Estimated => EdiQuantityQuality.Estimated,
+            QuantityQuality.Unspecified => EdiQuantityQuality
+                .Missing, //throw new InvalidOperationException("Quality is not specified"),
```
Comment (lasrinnil) [inline source/IntegrationTests/Infrastructure/InboxEvents/TestAggregatedTimeSeriesRequestAcceptedHandlerSpy.cs:71]: ?

File: source/IntegrationTests/Infrastructure/InboxEvents/TestAggregatedTimeSeriesRequestAcceptedHandlerSpy.cs [hunk 6]
```diff
@@ -59,15 +59,16 @@ public Task Handle(AggregatedTimeSerieRequestWasAccepted notification, Cancellat
...
-            QuantityQuality.Incomplete => Quality.Incomplete.Name,
-            QuantityQuality.Measured => Quality.Measured.Name,
-            QuantityQuality.Missing => Quality.Missing.Name,
-            QuantityQuality.Estimated => Quality.Estimated.Name,
-            QuantityQuality.Unspecified => throw new InvalidOperationException("Quality is not specified"),
+            QuantityQuality.Incomplete => EdiQuantityQuality.NotAvailable,
+            QuantityQuality.Measured => EdiQuantityQuality.Measured,
+            QuantityQuality.Missing => EdiQuantityQuality.Missing,
+            QuantityQuality.Estimated => EdiQuantityQuality.Estimated,
+            QuantityQuality.Unspecified => EdiQuantityQuality
```
Comment (lasrinnil) [inline source/IntegrationTests/Infrastructure/InboxEvents/TestAggregatedTimeSeriesRequestAcceptedHandlerSpy.cs:70]: make sure this mapping is correct - maybe ask @JesperGomes

File: source/OutgoingMessages.Application/MarketDocuments/AggregationResult/AggregationResultXmlDocumentWriter.cs [hunk 7]
```diff
@@ -107,12 +107,19 @@ protected override async Task WriteMarketActivityRecordsAsync(IReadOnlyCollectio
 
     private Task WriteQualityIfRequiredAsync(XmlWriter writer, Point point)
     {
-        if (point.Quality is null)
-            return Task.CompletedTask;
+        // TODO (MOB): Hmm...
```
Comment (lasrinnil) [inline source/OutgoingMessages.Application/MarketDocuments/AggregationResult/AggregationResultXmlDocumentWriter.cs:110]: this is fine 🥷

File: source/Process.Application/Transactions/Aggregations/AggregatedTimeSeriesRequestAcceptedEventMapper.cs [hunk 8]
```diff
@@ -122,15 +122,16 @@ private static string MapUnitType(QuantityUnit quantityUnit)
...
-            QuantityQuality.Incomplete => Quality.Incomplete.Name,
-            QuantityQuality.Measured => Quality.Measured.Name,
-            QuantityQuality.Missing => Quality.Missing.Name,
-            QuantityQuality.Estimated => Quality.Estimated.Name,
-            QuantityQuality.Unspecified => throw new InvalidOperationException("Quality is not specified"),
+            QuantityQuality.Incomplete => EdiQuantityQuality.NotAvailable,
+            QuantityQuality.Measured => EdiQuantityQuality.Measured,
+            QuantityQuality.Missing => EdiQuantityQuality.Missing,
+            QuantityQuality.Estimated => EdiQuantityQuality.Estimated,
+            QuantityQuality.Unspecified => EdiQuantityQuality
```
Comment (lasrinnil) [inline source/Process.Application/Transactions/Aggregations/AggregatedTimeSeriesRequestAcceptedEventMapper.cs:133]: Verify this is correct, ask @JesperGomes perhaps 😄

File: source/Tests/Domain/OutgoingMessages/MarketDocuments/EbixCodeTests.cs [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+
+        // Assert
+        Assert.Equal(expectedResult, result);
+    }
+
+    [Theory]
+    [MemberData(nameof(EdiQualityValues))]
+    public void Can_handle_all_edi_qualities(EdiQuantityQuality ediQuantityQuality)
+    {
+        // Act
```
Comment (lasrinnil) [inline source/Tests/Domain/OutgoingMessages/MarketDocuments/EbixCodeTests.cs:52]:
```suggestion
        // Act & Assert
```

Comment (lasrinnil) [inline source/Tests/Domain/OutgoingMessages/MarketDocuments/EbixCodeTests.cs:52]: Or simply remove the comment

File: source/Tests/Domain/Transactions/Aggregations/EdiQualityMapperTests.cs [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+
+public sealed class EdiQualityMapperTests
+{
+#pragma warning disable CA1034
+    public sealed class QuantityQualityCollectionToEdiQuality
+#pragma warning restore CA1034
+    {
+        public static IEnumerable<object[]> QualityMappingData()
+        {
+            // Example mappings from the documentation
```
Comment (lasrinnil) [inline source/Tests/Domain/Transactions/Aggregations/EdiQualityMapperTests.cs:33]: provide a link or at least a sample of the documentation for reference

File: source/Tests/Domain/Transactions/Aggregations/EdiQualityMapperTests.cs [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+                {
+                    EnergyResultProducedV2.Types.QuantityQuality.Measured,
+                    EnergyResultProducedV2.Types.QuantityQuality.Measured,
+                    EnergyResultProducedV2.Types.QuantityQuality.Calculated,
+                },
+                EdiQuantityQuality.Measured,
+            };
+        }
+
+        public static IEnumerable<object[]> QuantityQualityValues()
```
Comment (lasrinnil) [inline source/Tests/Domain/Transactions/Aggregations/EdiQualityMapperTests.cs:145]: Why object? Cant this be typed?

File: source/Tests/Domain/Transactions/Aggregations/EdiQualityMapperTests.cs [hunk 12]
```diff
@@ -0,0 +1,183 @@
...
+            };
+
+            yield return new object[]
+            {
+                new[]
+                {
+                    EnergyResultProducedV2.Types.QuantityQuality.Missing,
+                    EnergyResultProducedV2.Types.QuantityQuality.Calculated,
+                },
+                EdiQuantityQuality.Missing,
```
Comment (lasrinnil) [inline source/Tests/Domain/Transactions/Aggregations/EdiQualityMapperTests.cs:102]:
```suggestion
            yield return new object[]
            {
                new[] // input
                {
                    EnergyResultProducedV2.Types.QuantityQuality.Missing,
                    EnergyResultProducedV2.Types.QuantityQuality.Calculated,
                },
                EdiQuantityQuality.Missing, // expected
```

File: source/Tests/Domain/Transactions/Aggregations/EdiQualityMapperTests.cs [hunk 13]
```diff
@@ -0,0 +1,183 @@
...
+            {
+                new[] { EnergyResultProducedV2.Types.QuantityQuality.Measured }, EdiQuantityQuality.Measured,
+            };
+
+            yield return new object[]
+            {
+                new[] { EnergyResultProducedV2.Types.QuantityQuality.Calculated }, EdiQuantityQuality.Calculated,
+            };
+
+            yield return new object[]
```
Comment (lasrinnil) [inline source/Tests/Domain/Transactions/Aggregations/EdiQualityMapperTests.cs:95]: Commenting what is the input and the expected value would help get a quicker grasp of the tests.

File: source/Process.Application/Transactions/Aggregations/EdiQualityMapper.cs [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+// limitations under the License.
+
+using System;
+using System.Collections.Generic;
+using Energinet.DataHub.EDI.BuildingBlocks.Domain.Models;
+using Energinet.DataHub.Wholesale.Contracts.IntegrationEvents;
+
+namespace Energinet.DataHub.EDI.Process.Application.Transactions.Aggregations;
+
+public static class EdiQualityMapper
```
Comment (lasrinnil) [inline source/Process.Application/Transactions/Aggregations/EdiQualityMapper.cs:22]: Please provide a quick summary of how this mapper is expected to work, that way if another person has to refactor, they will have the code, the tests, and the human-readable documentation provided.

File: source/OutgoingMessages.Domain/MarketDocuments/CimCode.cs [hunk 15]
```diff
@@ -149,22 +149,18 @@ public static string Of(Resolution resolution)
...
-            return "A06";
-        if (quality == Quality.Measured)
-            return "A04";
-        if (quality == Quality.Missing)
-            return "A02";
-
-        throw NoCodeFoundFor(quality.Name);
+        return calculatedQuantityQuality switch
+        {
+            CalculatedQuantityQuality.Missing => "A05",
```
Comment (lasrinnil) [inline source/OutgoingMessages.Domain/MarketDocuments/CimCode.cs:156]: Lets remove all these magic strings and instead use static strings within this class. An example would be: `CimCode.Missing => "A05"`

File: source/Process.Application/Transactions/Aggregations/AggregatedTimeSeriesRequestAcceptedEventMapper.cs [hunk 16]
```diff
@@ -122,14 +122,15 @@ private static string MapUnitType(QuantityUnit quantityUnit)
...
     {
         return quality switch
         {
-            QuantityQuality.Incomplete => Quality.Incomplete.Name,
-            QuantityQuality.Measured => Quality.Measured.Name,
-            QuantityQuality.Missing => Quality.Missing.Name,
-            QuantityQuality.Estimated => Quality.Estimated.Name,
+            QuantityQuality.Incomplete => CalculatedQuantityQuality.Incomplete,
+            QuantityQuality.Measured => CalculatedQuantityQuality.Measured,
+            QuantityQuality.Missing => CalculatedQuantityQuality.NotAvailable,
```
Comment (lasrinnil) [inline source/Process.Application/Transactions/Aggregations/AggregatedTimeSeriesRequestAcceptedEventMapper.cs:131]: Is it intended that QuantityQuality.Missing => CalculatedQuantityQuality.NotAvailable ?

File: source/Tests/Infrastructure/OutgoingMessages/AggregationResult/AggregationResultDocumentWriterTests.cs [hunk 17]
```diff
@@ -99,28 +107,86 @@ public async Task Point_quantity_element_is_excluded_if_no_value(string document
...
+            .ToList();
+
+        points.ForEach(point => _timeSeries.WithPoint(point));
+
+        var document = await CreateDocument(_timeSeries, DocumentFormat.From(documentFormat));
+
+        // Assert
+        points.Should().HaveCount(5);
+        AssertDocument(document, DocumentFormat.From(documentFormat))
+            .QualityIsPresentForPosition(1, "A05")
```
Comment (lasrinnil) [inline source/Tests/Infrastructure/OutgoingMessages/AggregationResult/AggregationResultDocumentWriterTests.cs:141]: Would be nice if these values were exposed by `CimCode`

File: source/Process.Application/Transactions/Aggregations/AggregatedTimeSeriesRequestAcceptedEventMapper.cs (hunk 16 above)
Comment (MWO1024) [reply source/Process.Application/Transactions/Aggregations/AggregatedTimeSeriesRequestAcceptedEventMapper.cs:131]: Indeed. I looked at the old code, at realised I made a mistake. We had:
`QQ.Incomplete => Old.Incomplete
QQ.Measued => Old.Measured
QQ.Missing => Old.Missing
QQ.Estimated => Old.Estimated
QQ._ => Error`
and
`Old.Estimated => A03
Old.Incomplete => A05
Old.Calculated => A06
Old.Measured => A04
Old.Missing => A02`
thus
`QQ.Missing => Old.Missing => A02` where A02 is "not available (used for resetting).

Conversely, as we now have
`New.Missing => A05
New.Incomplete => A05
New.Estimated => A03
New.Measured => A04
New.Calculated => A06
New.NotAvailable => A02`
we need
`QQ.Incomplete => New.Incomplete
QQ.Measured => New.Measured
QQ.Missing => New.NotAvailable
QQ.Estimated => New.Estimated
QQ.Calculated => New.Calculated
QQ._ => Error`

I have tried to summarise the quantity quality values (using Cim) here:
`N/A           => New.Missing      => A05 <= Old.Incomplete <= QQ.Incomplete
QQ.Incomplete => New.Incomplete   => A05 <= Old.Incomplete <= QQ.Incomplete
QQ.Estimated  => New.Estimated    => A03 <= Old.Estimated  <= QQ.Estimated
QQ.Measured   => New.Measured     => A04 <= Old.Measured   <= QQ.Measued
QQ.Calculated => New.Calculated   => A06 <= Old.Calculated <= N/A
QQ.Missing    => New.NotAvailable => A02 <= Old.Missing    <= QQ.Missing`

Comment (MWO1024) [reply source/Process.Application/Transactions/Aggregations/AggregatedTimeSeriesRequestAcceptedEventMapper.cs:131]: On the bright side, this potential confusion should disappear once we update the AggregatedTimeSeriesRequestAccepted event to also utilise a set of quantity qualities too. As long as the enum values of the two sets are isomorphic (which they should be), we can trivially create an bijection between them, and the translation from the two types of quantity quality sets to the new CalculatedQuantityQuality (and its invers) should be quite simple.

File: source/OutgoingMessages.Domain/MarketDocuments/EbixCode.cs [hunk 18]
```diff
@@ -19,6 +19,10 @@ namespace Energinet.DataHub.EDI.OutgoingMessages.Domain.MarketDocuments;
 
 public static class EbixCode
 {
+    public const string QuantityQualityCodeMeasured = "E01";
```
Comment (lasrinnil) [inline source/OutgoingMessages.Domain/MarketDocuments/EbixCode.cs:22]: with this change - now two patterns exist for values within `EbixCode` - consider creating a new PR uniforming them :)

**Review by lasrinnil — APPROVED** [review]

**Pull request merged**